### PR TITLE
inkscape/silhouette: init at 1.28

### PR DESCRIPTION
**Pull request by jfly**
###### Description of changes

Introduce the 'silhouette' extension for printing to Silhouette vinyl
cutters in Inkscape.

Tested with `nix-build -E "with import ./. { }; inkscape-with-extensions.override { inkscapeExtensions = [ inkscape-extensions.silhouette ]; }"`.

This completes https://github.com/NixOS/nixpkgs/issues/224852

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (raboof) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:21]: This may not be my personal favorite way to denote snapshot versions, but as there is [no consensus](https://github.com/NixOS/nixpkgs/issues/68518) on the 'correct way' IMHO this is as good as any :+1:

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:58]: It would probably be saner to just use `PREFIX` and patch udevdir to use PREFIX (ideally upstream). See https://github.com/NixOS/nixpkgs/issues/65718. Then we could use the default `installPhase` and pass the path with:

```nix
makeFlags = [
  "PREFIX=${placeholder "out"}"
];
```

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:17]: Could you add a comment why this is necessary? If using the original `.py` files with shell does not work for some reason.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:21]: The current guidelines recommend `1.27-unstable-2023-06-03`: https://nixos.org/manual/nixpkgs/unstable/#sec-package-naming

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (raboof) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:53]: Does it make sense to also include a `pythonImportsCheck` section here?

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:62]: Might be cleaner to move these to `postInstall`.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:74]: `gpl2` is unclear, use `gpl2Plus` or `gpl2Only`. See https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206 for more details.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:72]: Inkscape should be capitalised.

**Review by raboof — COMMENTED** [review]
On this commit (0ab50cfea23ff108aee2888125f5b9187af8a220) it shows the configuration pop up fine, I did not see any tkinter problem (#238990). Not tested with an actual silhouette though.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:50]: By convention, Python packages place this attribute after `version`.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:53]: I've never heard of `pythonImportsCheck`, so I don't know. Doing some reading, it sounds like it's a tool for [sanity checking a python package that doesn't have unit tests](https://github.com/jfly/nixpkgs/blob/0ab50cfea23ff108aee2888125f5b9187af8a220/doc/languages-frameworks/python.section.md?plain=1#L726-L727)? But in this case, we *do* have unit tests, so my initial reaction is that it seems like there is not much value in this.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:21]: Ooh, I'm not sure where I got this from, but I was following some other package in nixpkgs.

[According to @jtojnar](https://github.com/jfly/nixpkgs/blob/0ab50cfea23ff108aee2888125f5b9187af8a220/doc/languages-frameworks/python.section.md?plain=1#L726-L727), there *is* a recommendation, so I'll update this to follow that!

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:58]: Neat idea! A couple of thoughts:

1. I've added a patch to get their `Makefile` to use `PREFIX` and changed things to just use the default `installPhase`. However, I'm not optimistic about upstreaming it (because it would change the default value for `UDEV` from `/lib/udev` to `/usr/lib/udev`. Did you have something else in mind?
2. In order to use the default `installPhase` with mk-python-derivation.nix, it looks like I have to either set `format = "other";`, or specify `dontUsePipInstall = true;`. I opted for the `dontUsePipInstall` approach.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:17]: Oh thanks for pointing this out. This is definitely no obvious. Commented added!

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: Oops, I'm not sure how that slipped through the cracks, sorry! Changed to `gpl2Plus`.

Comment (raboof) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: How did you determine the `Plus`? I don't see any "or later" in https://github.com/search?q=repo%3Afablabnbg%2Finkscape-silhouette%20GPL&type=code

Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: I don't know much about software licenses, but I got this from here https://github.com/fablabnbg/inkscape-silhouette/blob/a1941b6f740466a20dfa116415dd599b1e496302/LICENSE#L298-L299 (emphasis mine):

> either version 2 of the License, or (at your option) any later version.

Comment (raboof) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: That section is part of the "How to Apply These Terms to Your New Programs" documentation for the license, but I don't see the project actually _do_ that AFAICS

Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: Ahh, I see. Then I really have no idea what the license is for this software. It doesn't seem to follow the GPL advice to include a notice on each file. Per the instructions [here](https://discourse.nixos.org/t/lib-licenses-gpl3-co-are-now-deprecated/8206), I've reached out to the maintainer [here](https://github.com/fablabnbg/inkscape-silhouette/issues/239) to see clarity.

What does this mean for this PR? Do we have to wait until we have a clear answer on this, or can we use the ambiguous `gpl2` until we get an answer on that github issue I posted?

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: @jtojnar I'm going to mark this discussion resolved, but feel free to reopen it if you're not happy with the decisions I made!

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jnweiger) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: The code is effectively GPLv2 only.

- setup.py:license='GPL-2.0',
- silhouette/Graphtec.py:# Distribute under GPLv2 or ask.
- misc/test_drive.py:# (c) 2013 jnweiger@gmail.com, distribute under GPL-2.0
- silhouette/convert2dashes.py: [GPL] either version 2 of the License, or (at your option) any later version.
- sendto_silhouette.py:# (C) 2013 jw@suse.de. Licensed under CC-BY-SA-3.0 or GPL-2.0 at your choice.
- silhouette-udev-notify.sh:# (c) 2013, jw@suse.de - All rights reserved. Distribute under GPL-2.0


While the main file sendto_silhouette.py is under a dual license we need to find what applies to the entire package. CC BY-SA-3.0 is not mentioned anywhere else, so we have at least 3 files that only allow GPLv2.

Regarding the hint in Graphtec.y: "Distribute under GPLv2 or ask." - I probably won't object to allow "GPLv2 or later", if someone asks for it.
But: by now the code base has contributions from many different authors, so my vote is insufficient.

I understand that the text of GPLv2 itself contains the suggestion to use it as "or later" https://github.com/fablabnbg/inkscape-silhouette/blob/main/LICENSE#L299
But this package did not follow that suggestion.

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jnweiger) [inline pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:79]: please change this to before merging:

```suggestion
    license = licenses.gpl2Only;
```

**Review by jnweiger — CHANGES_REQUESTED** [review]
it's GPL-2.0-Only

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix:79]: Done!

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: Thank you for the clarification, @jnweiger!

**Comment by jfly** [comment]
@raboof @jtojnar, I would love if you could take another look at this PR sometime. Thanks!

**Comment by raboof** [comment]
Looking at the docs at https://github.com/fablabnbg/inkscape-silhouette/blob/main/USERGUIDE.md , I don't see the "Extention > Render > Silhouette Regmarks" option - is that a problem with the packaging or an upstream problem?

File: pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix
Comment (jfly) [reply pkgs/applications/graphics/inkscape/extensions/silhouette/default.nix]: For the record, using `dontUsePipInstall` stopped working due to https://github.com/nixos/nixpkgs/commit/6c85fff302615c62bf4f632bca661bc48298b0a3, which introduces a new pypaInstallHook without a corresponding dont* option.

It still feels weird to me to change the format to "other" when I know this is a setuptools project, so I've opted to override the buildPhase to call `make` again.

Happy to tackle this some other way if you prefer.

**Comment by jfly** [comment]
> Looking at the docs at https://github.com/fablabnbg/inkscape-silhouette/blob/main/USERGUIDE.md , I don't see the "Extention > Render > Silhouette Regmarks" option - is that a problem with the packaging or an upstream problem?

Sorry for the delayed response here. The answer is neither. The "Silhouette Regmarks" option was added in https://github.com/fablabnbg/inkscape-silhouette/commit/140400e2b61c600abea539188c5649dea0de5d24#diff-bbada293bbff7deda32bb6e7e2574e6e26b15185f5dcb1ac8ba02ea090bbac06, which did not land in 1.28 (and there has not been a newer release yet at time of writing (2023-11-19)).

**Comment by jfly** [comment]
I've rebased this PR (and fixed [one issue that arose due to changes in the nixpkgs Python machinery in the mean time](https://github.com/NixOS/nixpkgs/pull/239349#discussion_r1398508827)). AFAIK, there are no outstanding issues with this PR, and I'd love to get it merged up if possible: @raboof @jtojnar.

**Review by raboof — APPROVED** [review]
not tested with an *actual* silhouette, but looks good as far as I've seen. sorry for the long delay in reviewing :)

**Pull request merged**
